zrml/court/src/lib.rs

```diff
@@ -206,31 +346,43 @@ mod pallet {
         type MarketId = MarketIdOf<T>;
 
         fn on_dispute(
+            dispute_bound: Self::Balance,
             disputes: &[MarketDispute<Self::AccountId, Self::BlockNumber>],
-            market_id: Self::MarketId,
+            market_id: &Self::MarketId,
+            who: &Self::AccountId,
         ) -> DispatchResult {
+            CurrencyOf::<T>::reserve(who, dispute_bound)?;
             let jurors: Vec<_> = Jurors::<T>::iter().collect();
             let necessary_jurors_num = Self::necessary_jurors_num(disputes);
             let mut rng = Self::rng();
             let random_jurors = Self::random_jurors(&jurors, necessary_jurors_num, &mut rng);
             let curr_block_num = <frame_system::Pallet<T>>::block_number();
             let block_limit = curr_block_num.saturating_add(T::CourtCaseDuration::get());
             for (ai, _) in random_jurors {
-                RequestedJurors::<T>::insert(ai, market_id, block_limit);
+                RequestedJurors::<T>::insert(market_id, ai, block_limit);
             }
             Ok(())
         }
 
         fn on_resolution<D>(
             _: &D,
             _: &[MarketDispute<Self::AccountId, Self::BlockNumber>],
-            _: &Self::MarketId,
+            market_id: &Self::MarketId,
             _: &Market<Self::AccountId, Self::BlockNumber>,
         ) -> Result<OutcomeReport, DispatchError>
         where
             D: Fn(usize) -> Self::Balance,
         {
-            Ok(OutcomeReport::Scalar(Default::default()))
+            let mut requested_jurors: Vec<_> =
+                RequestedJurors::<T>::iter_prefix(market_id).collect();
+            let votes: Vec<_> = Votes::<T>::iter_prefix(market_id).collect();
+            Self::punish_tardy_jurors(&mut requested_jurors)?;
+            let (first, second) = Self::two_best_outcomes(&votes)?;
+            Self::set_jurors_that_sided_on_the_second_most_voted_outcome_as_tardy(&second, &votes)?;
+            Self::set_late_jurors_as_tardy(&requested_jurors, &votes)?;
+            Votes::<T>::remove_prefix(market_id, None);
+            RequestedJurors::<T>::remove_prefix(market_id, None);
+            Ok(first)
```

Comment:
> The distribution of the funds (losers -> winners) will be done in an upcoming PR, right?
Yes, it was implemented on https://github.com/zeitgeistpm/zeitgeist/pull/276

> Something different: Currently a juror marked as tardy is doomed to be slashed and removed during the second vote phase, irrespective of what the juror does. This is because on the first misbehavior, the juror will be marked tardy, the tardy flag will never be removed though, consequently the juror will be punished during the next on_resolution phase in which the juror is part of.

Oopss.... My bad.
Thanks for spotting this problem